### ISSUE TITLE
[css-pseudo] Define ::backdrop

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#7845]
Right now `::backdrop` is defined in https://fullscreen.spec.whatwg.org/#::backdrop-pseudo-element

> Each element in a [top layer](https://fullscreen.spec.whatwg.org/#top-layer) has a ::backdrop pseudo-element. This pseudo-element is a box rendered immediately below the element (and above the element before the element in the set, if any), within the same [top layer](https://fullscreen.spec.whatwg.org/#top-layer).
> It does not inherit from any element and is not inherited from. No restrictions are made on what properties apply to this pseudo-element either. 

Well, there are several details that are not clear to me:
 - Is it a tree-abiding pseudo-element?
 - If it's a tree-abiding pseudo-element, where is it originated? Before `::marker`, between `::marker` and `::before`, between `::before` and 1st child, between last child and `::after`, after `::after`?
 - It accepts random properties, e.g. `display: list-item`. But then:
    - Does it really become a list item?
    - If it becomes a list item, does it increment the implicit `list-item` counter as usual? If so, it's position in the tree matters.
    - Do counter properties even work in `::backdrop`? The properties are defined as applying to "all elements (includes ::before and ::after pseudo-elements)", but the `::backdrop` definition says "no restrictions".
    - Does it originate a `::marker` pseudo-element?
    - Should `::backdrop::marker` be a valid selector?

I think css-pseudo should define `::backdrop` and clarify these details.

**Comment by bramus** [comment]
Additionally: the [current spec](https://fullscreen.spec.whatwg.org/#::backdrop-pseudo-element) reads _“It does not inherit from any element and is not inherited from.”_

This also means that custom properties are not available inside of `::backdrop`, as demonstrated through [this demo](https://codepen.io/mattwilcoxuk/pen/abKbpMj):

```css
:root {
  --brand: 0 50% 40%;
}

dialog {
  background-color: hsl( var(--brand) / 0.5 );
}

dialog::backdrop {
  background-color: hsl( var(--brand) / 0.5 ); /* ❌ Does not work */
  backdrop-filter: blur(0.2rem);
}
```

This [confuses authors](https://twitter.com/MattWilcox/status/1584847908062593030?s=20&t=ELtOhQrV9NpFSmY0agziIw) as they – and me included as well – expected custom props to be available there, similar to them being available in `::after` for example.

My proposal for this specific detail, would be to allow custom properties to inherit down into `::backdrop`.

**Comment by Loirooriol** [comment]
Changing this behavior should have its own issue, now for the fullscreen spec there is https://github.com/whatwg/fullscreen/issues/124

**Comment by Loirooriol** [comment]
@tabatkins In https://drafts.csswg.org/css-position-4/#backdrop you defined `::backdrop` as tree-abiding.
Then the questions from the 1st comment need to be answered.
In https://chromium-review.googlesource.com/c/chromium/src/+/3933059 I assumed that `::backdrop` is not tree-abiding and can't be a list item.

**Comment by tabatkins** [comment]
Ah, thanks for this ping! My initial thoughts:

> Is it a tree-abiding pseudo-element?

Yes, in that it's a well-defined box with a definition location, so should accept all the stuff other boxes do.

> If it's a tree-abiding pseudo-element, where is it originated? Before ::marker, between ::marker and ::before, between ::before and 1st child, between last child and ::after, after ::after?

None of these; it is generated in the top layer, as an independent box with no particular relationship to its originating element. (But it's rendered just below the originating element.)

> It accepts random properties, e.g. display: list-item. But then:
> Does it really become a list item?

No opinion. I'm fine with forcing it to be `display: block` via a UA style rule.

> If it becomes a list item, does it increment the implicit list-item counter as usual? If so, it's position in the tree matters.

Right, so having this be impossible is probably reasonable.

> Do counter properties even work in ::backdrop? The properties are defined as applying to "all elements (includes ::before and ::after pseudo-elements)", but the ::backdrop definition says "no restrictions".

I suppose they do, but we will indeed need to define the position in the tree.

I propose that they're an isolated context that doesn't see any other counters, and its own counters are also invisible to any other element.

> Does it originate a ::marker pseudo-element?

If we force it to `display: block`, no. If not, yes.

> Should ::backdrop::marker be a valid selector?

It should probably be *valid* regardless of the answer to the previous question, but I suspect we'll end up having it not be capable of matching anything since you won't be able to make it a list item.

**Comment by Crissov** [comment]
> Should `::backdrop::marker` be a valid selector?

No. 

Not so sure about `::marker::backdrop` however.

**Comment by tabatkins** [comment]
`::marker::backdrop` can't ever exist, the marker pseudo can't be put in the top layer.

**Comment by tabatkins** [comment]
Nothing wrong with a tree having many roots. It's still tree-abiding, as in, it has a definite set of boxes with definite locations and boundaries, which do not violate the existing tree-ness of the element/box tree (unlike ::first-line).

**Comment by Crissov** [comment]
> Each element in the top layer has a ::backdrop pseudo-element, for which it is the originating element.

Just to clarify, this means there is no way whatsoever to reveal a box’s backdrop unless the originating element is in the top layer. 

I’d find this more generic UA default stylesheet interesting:

~~~~ css
::backdrop {
  display: none;
}
:fullscreen::backdrop {
  display: block;
}
~~~~

On the other hand, I’m wondering whether the use case for this pseudo-element couldn‘t have been covered by something like `background-origin: viewport`. 

On the third hand, I would have preferred if backgrounds had been done with pseudo-elements and the `fill` or the `content` property from the start. 

~~~~ css
foo::background(0) {
  content: url("bg.img");
  fill: navy;
}
~~~~

**Comment by Loirooriol** [comment]
@Crissov There is no need for that `display` style since pseudo-elements can be prevented from generating boxes by other means, e.g. `li::marker` will only be generated if the list item has `display: list-item`, and we don't need a `li:list-item::marker` rule with a circular selector.

**Comment by tabatkins** [comment]
Agenda+ to discuss a few questions about ::backdrop's display and position in the tree:

1. Should we just force ::backdrop to `display: block !important;` in the UA stylesheet? (This avoids some follow-up questions.)
2. Is ::backdrop indeed tree-abiding, or should we lock it down more?
	* Does it have a ::before/::after?
	* Is it fine for any property to work on it?
3. How should ::backdrop act wrt various tree/sibling-scoped things, like counters?


My preferences are:
1. Yes, ::backdrop doesn't have contents and only exists to hold a color or an image and maybe block mouse events, so there's no need to deal with other display values.
2. Yes, it's tree-abiding. No opinion on ::before/::after; I'm fine if we explicitly say it doesn't generate those. 
3. I think the easiest is to say that each backdrop is an independent tree for these purposes, not related to the document tree. It doesn't see counters from the document, and the document doesn't see counters from it; etc.  Alternately, backdrop can be a child of the originating element (maybe after the ::after pseudo?); this might be more straightforward, since it does inherit from the originating element now.

**Comment by bramus** [comment]
1. Yes
2. Tree-abiding: yes. before/after: fine with locking it down.
3. I see `::backdrop` as a child of its originating element. It just happens to render in the top-layer.

**Comment by brandonmcconnell** [comment]
1. I see the rationale in using `display: block !important` (w/ `!important`). If we're going to enforce the `display: block` on the UA, should we consider enforcing other properties that could be used to work around that limitation, like `visibility: hidden; pointer-events: none;`?
2. I would think tree-abiding too; before/after could go either way, but I don't see any strong cases reasons to prevent using them
3. Agreed with all the general thoughts so far, namely @bramus' take— "`::backdrop` as a child of its originating element. It just happens to render in the top-layer."

**Comment by tabatkins** [comment]
No, making the backdrop `pointer-events:none` is something we def want to support. We're not trying to prevent it from being hidden, just limiting what sort of complexity we can potentially get into for what is always going to be an empty element. (That is, `display:none` would be okay; it's `display:grid` or `display:list-item` we want to avoid if possible.)

**Comment by brandonmcconnell** [comment]
Sweet, I like that approach! I could definitely see hiding it sometimes.

**Comment by Loirooriol** [comment]
I'm against `display: none !important` in UA because I think `display: none` makes sense. I would be fine forcing non-`none` (and maybe non-`contents`) values compute to `block` in the style adjuster.

**Comment by Loirooriol** [comment]
I didn't realize this during the call, but it's possible to have multiple elements in the top layer, each one with its own ::backdrop. Then, if all these ::backdrop are siblings of the root, what's their relative order? Are they ordered e.g. according to the tree order of their originating elements?
